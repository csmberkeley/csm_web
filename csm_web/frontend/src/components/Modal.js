import React, { useContext } from "react";
import PropTypes from "prop-types";

const ModalContext = React.createContext(Function.prototype);

export default function Modal({ children, closeModal, className = "" }) {
  return (
    <ModalContext.Provider value={closeModal}>
      <div className="modal-overlay" onClick={closeModal} />
      <div className={`modal ${className}`}>
        <div className="modal-contents">
          <div className="modal-close-x">
            <button className="inline-plus-sign" onClick={closeModal}>
              <span>+</span>
            </button>
          </div>
          {children}
        </div>
      </div>
    </ModalContext.Provider>
  );
}

Modal.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]).isRequired,
  closeModal: PropTypes.func.isRequired,
  className: PropTypes.string
};

export function ModalCloser({ children }) {
  const closeModal = useContext(ModalContext);
  return (
    <div onClick={closeModal} className="modal-close">
      {children}
    </div>
  );
}

ModalCloser.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]).isRequired
};
