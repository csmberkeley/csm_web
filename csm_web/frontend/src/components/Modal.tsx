import React, { useContext } from "react";

const ModalContext = React.createContext(Function.prototype);

interface ModalProps {
  children: React.ReactNode;
  closeModal: React.MouseEventHandler<Element>;
  className?: string;
}

interface ModalCloserProps {
  children: React.ReactNode;
}

export default function Modal({ children, closeModal, className = "" }: ModalProps): React.ReactElement {
  return (
    <ModalContext.Provider value={closeModal}>
      <div className="modal-overlay" onClick={closeModal} />
      <div className={`modal ${className}`}>
        <div className="modal-contents">
          <div className="modal-close-x">
            <button className="inline-plus-sign" onClick={closeModal}>
              <span>Ã—</span>
            </button>
          </div>
          {children}
        </div>
      </div>
    </ModalContext.Provider>
  );
}

export function ModalCloser({ children }: ModalCloserProps): React.ReactElement {
  const closeModal = useContext(ModalContext) as React.MouseEventHandler<Element>;
  return (
    <div onClick={closeModal} className="modal-close">
      {children}
    </div>
  );
}
