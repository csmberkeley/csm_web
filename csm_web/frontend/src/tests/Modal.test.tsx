import React from "react";
import renderer from "react-test-renderer";
import Modal, { ModalCloser } from "../components/Modal";

/**
 * Returns a test Modal component.
 */
const getTestModal = (handleCloseModal: React.MouseEventHandler<Element>): React.ReactElement => {
  return (
    <Modal closeModal={handleCloseModal} className="test-modal">
      <div>Hello World</div>
    </Modal>
  );
};

describe("Modal", () => {
  it("should render correctly", () => {
    const handleCloseModal = jest.fn();
    const component = renderer.create(getTestModal(handleCloseModal));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("should close correctly", () => {
    test("when the close button is clicked", () => {
      const handleCloseModal = jest.fn();
      const component = renderer.create(getTestModal(handleCloseModal));
      const root = component.root;

      // find and click the close button
      const closeButton = root.findByProps({ className: "modal-close-x" }).findByType("button");
      closeButton.props.onClick();

      expect(handleCloseModal).toHaveBeenCalled();
    });

    test("when the overlay is clicked", () => {
      const handleCloseModal = jest.fn();
      const component = renderer.create(getTestModal(handleCloseModal));
      const root = component.root;

      // find and click the overlay
      const overlay = root.findByProps({ className: "modal-overlay" });
      overlay.props.onClick();

      expect(handleCloseModal).toHaveBeenCalled();
    });

    test("when an external modal closer is clicked", () => {
      // create a modal
      const handleCloseModal = jest.fn();
      const component = renderer.create(
        <Modal closeModal={handleCloseModal} className="test-modal">
          <div>Hello World</div>
          <ModalCloser>Close</ModalCloser>
        </Modal>
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      // find and click the external modal closer
      const root = component.root;
      const closer = root.findByProps({ className: "modal-close" });
      closer.props.onClick();

      expect(handleCloseModal).toHaveBeenCalled();
    });
  });
});
