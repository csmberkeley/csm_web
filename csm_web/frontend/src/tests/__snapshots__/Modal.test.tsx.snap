// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should close correctly when an external modal closer is clicked 1`] = `
Array [
  <div
    className="modal-overlay"
    onClick={[MockFunction]}
  />,
  <div
    className="modal test-modal"
  >
    <div
      className="modal-contents"
    >
      <div
        className="modal-close-x"
      >
        <button
          className="inline-plus-sign"
          onClick={[MockFunction]}
        >
          <span>
            ×
          </span>
        </button>
      </div>
      <div>
        Hello World
      </div>
      <div
        className="modal-close"
        onClick={[MockFunction]}
      >
        Close
      </div>
    </div>
  </div>,
]
`;

exports[`Modal should render correctly 1`] = `
Array [
  <div
    className="modal-overlay"
    onClick={[MockFunction]}
  />,
  <div
    className="modal test-modal"
  >
    <div
      className="modal-contents"
    >
      <div
        className="modal-close-x"
      >
        <button
          className="inline-plus-sign"
          onClick={[MockFunction]}
        >
          <span>
            ×
          </span>
        </button>
      </div>
      <div>
        Hello World
      </div>
    </div>
  </div>,
]
`;
